### ISSUE TITLE
Unexpected crash when inferring attribute on node

### ISSUE THREAD

**Issue by tushar-deepsource** [PyCQA/pylint#5479]
### Bug description

It's hard to reproduce this bug. But it happened when I was trying to run pylint over [this repository](https://github.com/School-Simplified/Timmy-SchoolSimplified)

Here's the logs, they should be fairly helpful in tracking down the issue:

```log
Exception on node <AssignAttr.read_messages l.1027 at 0x11ed3bf70> in file '/tmp/Timmy-SchoolSimplified/utils/events/TicketDropdown.py'
Traceback (most recent call last):
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 75, in walk
    callback(astroid)
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/pylint/checkers/classes.py", line 1459, in visit_assignattr
    self._check_in_slots(node)
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/pylint/checkers/classes.py", line 1511, in _check_in_slots
    if node.attrname in klass.locals and _has_data_descriptor(
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/pylint/checkers/classes.py", line 385, in _has_data_descriptor
    attributes = cls.getattr(attr)
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes.py", line 2596, in getattr
    raise AttributeInferenceError(
astroid.exceptions.AttributeInferenceError: 'read_messages' not found on <ClassDef.PermissionOverwrite l.602 at 0x112e2d720>.
```

### Command used

```shell
pylint [spaces separated list of all python files in the repo]
```

### Pylint output

```shell
Exception on node <AssignAttr.read_messages l.1027 at 0x11ed3bf70> in file '/tmp/Timmy-SchoolSimplified/utils/events/TicketDropdown.py'
Traceback (most recent call last):
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 75, in walk
    callback(astroid)
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/pylint/checkers/classes.py", line 1459, in visit_assignattr
    self._check_in_slots(node)
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/pylint/checkers/classes.py", line 1511, in _check_in_slots
    if node.attrname in klass.locals and _has_data_descriptor(
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/pylint/checkers/classes.py", line 385, in _has_data_descriptor
    attributes = cls.getattr(attr)
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes.py", line 2596, in getattr
    raise AttributeInferenceError(
astroid.exceptions.AttributeInferenceError: 'read_messages' not found on <ClassDef.PermissionOverwrite l.602 at 0x112e2d720>.
```

### Expected behavior

No crashes :)

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.6
Python 3.10.0 (default, Oct 13 2021, 06:45:00) [Clang 13.0.0 (clang-1300.0.29.3)]

Same thing happens on pylint 2.12 with astroid 2.9.0 as well.
```

### OS / Environment

MacOS Monterey

### Additional dependencies

These are the dependencies, but I don't think that reproducing this issue is necessary for figuring out what's going wrong. I think it's just an uncaught exception that needs to be caught.

```
aiohttp==3.8.1
astroid==2.8.6
async-timeout==4.0.1
attrs==21.2.0
bleach==4.1.0
blinker==1.4
boto3==1.20.20
cachetools==4.2.4
certifi==2021.10.8
chardet==4.0.0
charset-normalizer==2.0.9
chat-exporter==1.7.2
click==8.0.3
colorama==0.4.4
configcat-client==5.0.0
discord-sentry-reporting==2.0.4
docutils==0.18.1
dpytest==0.5.3
emoji==1.6.1
Flask==2.0.2
google-auth==2.3.3
google-auth-oauthlib==0.4.6
google-cloud-texttospeech==2.9.0
grapheme==0.6.0
gspread==5.0.0
gTTS==2.2.3
httplib2==0.20.2
idna==3.3
importlib-metadata==4.8.2
isort==5.10.1
itsdangerous==2.0.1
Jinja2==3.0.3
jishaku==2.3.2
keyring==23.4.0
lazy-object-proxy==1.6.0
logtail-python==0.1.2
MarkupSafe==2.0.1
mccabe==0.6.1
multidict==5.2.0
newrelic==7.2.4.171
oauth2client==4.1.3
oauthlib==3.1.1
packaging==21.3
peewee==3.14.8
pkginfo==1.8.2
platformdirs==2.4.0
psutil==5.8.0
py-cord @ git+https://github.com/Pycord-Development/pycord@9fceff873404ff57fcc1a3c3f07c539152226fbb
pyasn1==0.4.8
pyasn1-modules==0.2.8
Pygments==2.10.0
pylint==2.11.1
pylint-plugin-utils==0.6
PyMySQL==1.0.2
pyOpenSSL==21.0.0
pyparsing==3.0.6
pytest==6.2.5
python-dotenv==0.19.2
pytz==2021.3
readme-renderer==30.0
requests==2.26.0
requests-oauthlib==1.3.0
requests-toolbelt==0.9.1
rfc3986==1.5.0
rsa==4.8
six==1.16.0
tdqm==0.0.1
toml==0.10.2
tqdm==4.62.3
twine==3.7.0
typing_extensions==4.0.1
urllib3==1.26.7
webencodings==0.5.1
Werkzeug==2.0.2
wrapt==1.13.3
yarl==1.7.2
zipp==3.6.0
```

**Comment by DanielNoord** [comment]
A reproducible piece of code would definitely help in getting this solved. Just looking at:
`Exception on node <AssignAttr.read_messages l.1027 at 0x11ed3bf70> in file '/tmp/Timmy-SchoolSimplified/utils/events/TicketDropdown.py'` isn't that helpful cause L1027 doesn't include a `read_messages` reference.
Thus, anyone wanting to contribute to this will need to checkout the repository and find out what is actually causing the issue. That will slow down getting this solved 😉

**Comment by tushar-deepsource** [comment]
@DanielNoord My bad, here's the correct commit id where i reproduced this:

https://github.com/School-Simplified/Timmy-SchoolSimplified/blob/1c23a197f35ed57c13ed4e3ea668d6a892ddc92a/utils/events/TicketDropdown.py

**Comment by tushar-deepsource** [comment]
And just in case:
```console
$ git status          
HEAD detached at 1c23a19
Untracked files:
  (use "git add <file>..." to include in what will be committed)
        inp

nothing added to commit but untracked files present (use "git add" to track)
```

**Comment by tushar-deepsource** [comment]
Going more minimal now:
```console
$ pylint utils/event/TicketDropdown.py > /dev/null
xception on node <AssignAttr.read_messages l.1027 at 0x106f843d0> in file '/private/tmp/Timmy-SchoolSimplified/utils/events/TicketDropdown.py'
Traceback (most recent call last):
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 75, in walk
    callback(astroid)
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/pylint/checkers/classes.py", line 1459, in visit_assignattr
    self._check_in_slots(node)
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/pylint/checkers/classes.py", line 1511, in _check_in_slots
    if node.attrname in klass.locals and _has_data_descriptor(
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/pylint/checkers/classes.py", line 385, in _has_data_descriptor
    attributes = cls.getattr(attr)
  File "/private/tmp/Timmy-SchoolSimplified/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes.py", line 2596, in getattr
    raise AttributeInferenceError(
astroid.exceptions.AttributeInferenceError: 'read_messages' not found on <ClassDef.PermissionOverwrite l.602 at 0x109362530>.
utils/events/TicketDropdown.py:1:0: F0002: utils/events/TicketDropdown.py: Fatal error while checking 'utils/events/TicketDropdown.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/Users/tusharsadhwani/Library/Caches/pylint/pylint-crash-2021-12-06-16.txt'. (astroid-error)
```

**Comment by Wuestengecko** [comment]
Hello, I just happened to come across the same exception. Here's a minimal code example (just put this in a file and run pylint over it):

```python
from __future__ import annotations

import asyncio


class Connector:
    __slots__ = ("_Connector__reader", "_Connector__writer")

    __reader: asyncio.StreamReader
    __writer: asyncio.StreamWriter

    def __init__(self) -> None:
        raise TypeError("Use connect() instead")

    @classmethod
    async def connect(cls, socket: str) -> Connector:
        self = cls.__new__(cls)
        self.__reader, self.__writer = await asyncio.open_unix_connection(socket)
        return self


async def main():
    conn = await Connector.connect("/tmp/mysocket")


asyncio.run(main())

```

pylint crashed with a ``AttributeInferenceError`` and with the following stacktrace:

```
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/pylint/lint/pylinter.py", line 1034, in _check_files
    self._check_file(get_ast, check_astroid_module, file)
  File "/usr/lib/python3.9/site-packages/pylint/lint/pylinter.py", line 1069, in _check_file
    check_astroid_module(ast_node)
  File "/usr/lib/python3.9/site-packages/pylint/lint/pylinter.py", line 1203, in check_astroid_module
    retval = self._check_astroid_module(
  File "/usr/lib/python3.9/site-packages/pylint/lint/pylinter.py", line 1250, in _check_astroid_module
    walker.walk(node)
  File "/usr/lib/python3.9/site-packages/pylint/utils/ast_walker.py", line 75, in walk
    self.walk(child)
  File "/usr/lib/python3.9/site-packages/pylint/utils/ast_walker.py", line 75, in walk
    self.walk(child)
  File "/usr/lib/python3.9/site-packages/pylint/utils/ast_walker.py", line 75, in walk
    self.walk(child)
  [Previous line repeated 2 more times]
  File "/usr/lib/python3.9/site-packages/pylint/utils/ast_walker.py", line 72, in walk
    callback(astroid)
  File "/usr/lib/python3.9/site-packages/pylint/checkers/classes.py", line 1466, in visit_assignattr
    self._check_in_slots(node)
  File "/usr/lib/python3.9/site-packages/pylint/checkers/classes.py", line 1518, in _check_in_slots
    if node.attrname in klass.locals and _has_data_descriptor(
  File "/usr/lib/python3.9/site-packages/pylint/checkers/classes.py", line 344, in _has_data_descriptor
    attributes = cls.getattr(attr)
  File "/usr/lib/python3.9/site-packages/astroid/nodes/scoped_nodes.py", line 2744, in getattr
    raise AttributeInferenceError(
astroid.exceptions.AttributeInferenceError: '__reader' not found on <ClassDef.Connector l.6 at 0x7f8f16e8a9a0>.
```

It seems that in this specific case, the error is triggered by the combination of `__slots__`, the type annotations in the class body, and the variable names having double leading underscores.

You can make the crash disappear by doing any one of the following:

* Remove `__slots__`
* Delete the `__reader` and `__writer` annotations
* Move the annotations into `__init__` (without assigning anything to them)
* Use single leading underscores (i.e. `_reader` instead of `__reader`), and adjust the `__slots__` accordingly
* Delete the assignment to `__reader` and `__writer` in `connect()`

In the code referred to by @tushar-deepsource, I can neither find `__slots__` nor double leading underscores being used anywhere in the offending file, so it's very much possible that these are actually two separate error conditions producing the same symptoms. In that case, I'd be happy to file this as a new bug.

---

    % pylint --version
    pylint 2.12.2
    astroid 2.9.0
    Python 3.10.0 (default, Oct 29 2021, 16:38:21) [GCC 11.1.0]

**Comment by tushar-deepsource** [comment]
@Wuestengecko I'd assume that the error going away by doing these random things is because astroid inference has been really random, and inference stops working by seemingly unrelated changes. Because there's no inference going on, the buggy part of code is never hit and it results in no crash.

**Comment by DanielNoord** [comment]
Possible fix in #5494. I would love input on it, as I'm not actually sure we are now reporting `assigning-non-slot` correctly with that fix!

**Comment by DanielNoord** [comment]
Fixed with #5494.
